**Tugas Besar Machine Learning 2 : Classification**

Adiv Harjadinata - 1301183484

IF - 42 - 01

# Library

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score

# Dataset

## Dataset Overview

Mengimport dataset salju dan mengambil kolom yang akan diamati yaitu SuhuMin, SuhuMax, Hujan, SinarMatahari, Awan3pm, Suhu3pm, dan Bersalju besok. Berdasarkan panduan dalam mengerjakan dataset Salju

Kemudian menampilkan gambaran isi dari dataset

In [2]:
df_salju_test = pd.read_csv('salju_test.csv')
df_salju_test = df_salju_test[['SuhuMin','SuhuMax','Hujan','SinarMatahari','Awan3pm','Suhu3pm','BersaljuBesok']]
df_salju_test.head()

SuhuMin  SuhuMax  Hujan  SinarMatahari  Awan3pm  Suhu3pm BersaljuBesok
0     11.0     27.5    0.0            6.4      7.0     23.6            No
1     10.0     19.9    0.2            NaN      1.0     18.9           Yes
2      9.2     27.2    0.0           10.4      1.0     26.3            No
3     17.7     27.0    0.0            6.7      6.0     26.4            No
4      2.3      7.9   88.0            NaN      NaN      6.0            No

In [3]:
df_salju_train = pd.read_csv('salju_train.csv')
df_salju_train = df_salju_train[['SuhuMin','SuhuMax','Hujan','SinarMatahari','Awan3pm','Suhu3pm','BersaljuBesok']]
df_salju_train.head()

SuhuMin  SuhuMax  Hujan  SinarMatahari  Awan3pm  Suhu3pm BersaljuBesok
0     10.4     15.5    4.8            NaN      NaN     15.0         Tidak
1      9.0     17.0    8.0            7.4      5.0     15.5            Ya
2     18.2     32.0    0.0            NaN      NaN     29.6         Tidak
3      7.3     24.5    0.0           10.4      7.0     23.2         Tidak
4      5.9     20.3    0.0           12.6      6.0     18.1         Tidak

Menampilkan tipe dari kolom dataset

In [4]:
df_salju_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18182 entries, 0 to 18181
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SuhuMin        18017 non-null  float64
 1   SuhuMax        18017 non-null  float64
 2   Hujan          17795 non-null  float64
 3   SinarMatahari  9464 non-null   float64
 4   Awan3pm        10726 non-null  float64
 5   Suhu3pm        17740 non-null  float64
 6   BersaljuBesok  17763 non-null  object 
dtypes: float64(6), object(1)
memory usage: 994.5+ KB


In [5]:
df_salju_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 109095 entries, 0 to 109094
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   SuhuMin        107973 non-null  float64
 1   SuhuMax        108166 non-null  float64
 2   Hujan          106664 non-null  float64
 3   SinarMatahari  56716 non-null   float64
 4   Awan3pm        64624 non-null   float64
 5   Suhu3pm        106397 non-null  float64
 6   BersaljuBesok  106664 non-null  object 
dtypes: float64(6), object(1)
memory usage: 5.8+ MB


## Data Preprocessing

### Mengisi data kolom yang bernilai NaN

Melakukan perulangan dari dataset untuk mengisi nilai data dari setiap kolom yang berisi NaN dengan median dari kolom tersebut jika bertipe float64.
Jika bertipe object akan diisi dengan mean

In [6]:
for col in df_salju_test:
  if df_salju_test[col].dtype == 'float64':
    df_salju_test[col] = df_salju_test[col].fillna(df_salju_test[col].median())
  else:
    df_salju_test[col] = df_salju_test[col].fillna(df_salju_test[col].mode()[0])
    
df_salju_test

SuhuMin  SuhuMax  Hujan  SinarMatahari  Awan3pm  Suhu3pm BersaljuBesok
0         11.0     27.5    0.0            6.4      7.0     23.6            No
1         10.0     19.9    0.2            8.4      1.0     18.9           Yes
2          9.2     27.2    0.0           10.4      1.0     26.3            No
3         17.7     27.0    0.0            6.7      6.0     26.4            No
4          2.3      7.9   88.0            8.4      5.0      6.0            No
...        ...      ...    ...            ...      ...      ...           ...
18177      7.8     25.8    0.0           13.2      1.0     24.0            No
18178     12.4     26.6    0.0           11.6      1.0     25.7            No
18179      6.7     22.9    0.0            8.4      2.0     22.0            No
18180     12.5     26.7    0.0           10.8      5.0     25.4            No
18181     12.3     15.9    8.2            8.4      5.0     15.4           Yes

[18182 rows x 7 columns]

In [7]:
for col in df_salju_train:
  if df_salju_train[col].dtype == 'float64':
    df_salju_train[col] = df_salju_train[col].fillna(df_salju_train[col].median())
  else:
    df_salju_train[col] = df_salju_train[col].fillna(df_salju_train[col].mode()[0])
    
df_salju_train

SuhuMin  SuhuMax  Hujan  SinarMatahari  Awan3pm  Suhu3pm BersaljuBesok
0          10.4     15.5    4.8            8.4      5.0     15.0         Tidak
1           9.0     17.0    8.0            7.4      5.0     15.5            Ya
2          18.2     32.0    0.0            8.4      5.0     29.6         Tidak
3           7.3     24.5    0.0           10.4      7.0     23.2         Tidak
4           5.9     20.3    0.0           12.6      6.0     18.1         Tidak
...         ...      ...    ...            ...      ...      ...           ...
109090     20.1     23.7    0.0            8.9      6.0     22.1            Ya
109091     15.7     25.2    0.0            8.4      5.0     24.4            Ya
109092      7.5     20.4    1.6            8.4      5.0     19.8         Tidak
109093     10.8     29.8    0.0           11.2      1.0     29.2         Tidak
109094     12.3     27.4    9.0            8.4      5.0     23.3         Tidak

[109095 rows x 7 columns]

### Pengecekan Data NaN

Melakukan pengecekan kembali apakah masih ada nilai data yang berisi NaN

In [8]:
df_salju_test.isna().sum()

SuhuMin          0
SuhuMax          0
Hujan            0
SinarMatahari    0
Awan3pm          0
Suhu3pm          0
BersaljuBesok    0
dtype: int64

In [9]:
df_salju_train.isna().sum()

SuhuMin          0
SuhuMax          0
Hujan            0
SinarMatahari    0
Awan3pm          0
Suhu3pm          0
BersaljuBesok    0
dtype: int64

### Encode Label

Tahap ini dilakukan encoding label yaitu kolom BersaljuBesok.
Isi kolom ini akan diubah menjadi 1 dan 0 dari Ya dan Tidak.


Proses ini dilakukan agar saat memproses data tidak ada bertipe object dan full bertipe float64.

In [10]:
df_salju_test["BersaljuBesok"].replace(["Yes", "No"], [1,0], inplace = True)
df_salju_test

SuhuMin  SuhuMax  Hujan  SinarMatahari  Awan3pm  Suhu3pm  BersaljuBesok
0         11.0     27.5    0.0            6.4      7.0     23.6              0
1         10.0     19.9    0.2            8.4      1.0     18.9              1
2          9.2     27.2    0.0           10.4      1.0     26.3              0
3         17.7     27.0    0.0            6.7      6.0     26.4              0
4          2.3      7.9   88.0            8.4      5.0      6.0              0
...        ...      ...    ...            ...      ...      ...            ...
18177      7.8     25.8    0.0           13.2      1.0     24.0              0
18178     12.4     26.6    0.0           11.6      1.0     25.7              0
18179      6.7     22.9    0.0            8.4      2.0     22.0              0
18180     12.5     26.7    0.0           10.8      5.0     25.4              0
18181     12.3     15.9    8.2            8.4      5.0     15.4              1

[18182 rows x 7 columns]

In [11]:
df_salju_train["BersaljuBesok"].replace(["Ya", "Tidak"], [1,0], inplace = True)
df_salju_train

SuhuMin  SuhuMax  Hujan  SinarMatahari  Awan3pm  Suhu3pm  BersaljuBesok
0          10.4     15.5    4.8            8.4      5.0     15.0              0
1           9.0     17.0    8.0            7.4      5.0     15.5              1
2          18.2     32.0    0.0            8.4      5.0     29.6              0
3           7.3     24.5    0.0           10.4      7.0     23.2              0
4           5.9     20.3    0.0           12.6      6.0     18.1              0
...         ...      ...    ...            ...      ...      ...            ...
109090     20.1     23.7    0.0            8.9      6.0     22.1              1
109091     15.7     25.2    0.0            8.4      5.0     24.4              1
109092      7.5     20.4    1.6            8.4      5.0     19.8              0
109093     10.8     29.8    0.0           11.2      1.0     29.2              0
109094     12.3     27.4    9.0            8.4      5.0     23.3              0

[109095 rows x 7 columns]

Ekspor dataset yang sudah diolah

In [12]:
df_salju_test.to_csv('salju_test_terolah.csv')
df_salju_train.to_csv('salju_train_terolah.csv')

### Train Model

Melakukan split data menjadi dua variable yaitu x dan y

X adalah data keseluruhan
Y adalah data Label

In [13]:
x_test = df_salju_test.iloc[1:, :-1].values
y_test = df_salju_test.iloc[1:, -1].values

x_train = df_salju_train.iloc[1:, :-1].values
y_train = df_salju_train.iloc[1:, -1].values

### Scaling data

Scaling data dilakukan dengan **MinMaxScaler** agar persebaran data tidak terlalu jauh nilainya

In [14]:
min_max_scaler = MinMaxScaler()
x_test = min_max_scaler.fit_transform(x_test)
x_train = min_max_scaler.transform(x_train)

# Pemodelan

## Klasifikasi Data

Pada tahap klasifikasi dilakukan dengan metode **Logistic Regression** yang menjadi metode utama

In [15]:
model = LogisticRegression(random_state=0)
model.fit(x_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [16]:
predict = model.predict(x_test)
result = pd.concat([pd.DataFrame(y_test,columns=['Pasti']), pd.DataFrame(predict,columns=['Prediksi'])], axis=1)
result

Pasti  Prediksi
0          1         0
1          0         0
2          0         0
3          0         1
4          0         0
...      ...       ...
18176      0         0
18177      0         0
18178      0         0
18179      0         0
18180      1         0

[18181 rows x 2 columns]

# Evaluasi

Evaluasi dilakukan menggunakan **Classification Report, Confusion Matrix dan Accuracy Score**

In [17]:
from sklearn.metrics import classification_report

print(classification_report(y_test, predict, digits=4))

              precision    recall  f1-score   support

           0     0.8311    0.9560    0.8892     14242
           1     0.6518    0.2975    0.4086      3939

    accuracy                         0.8134     18181
   macro avg     0.7415    0.6268    0.6489     18181
weighted avg     0.7923    0.8134    0.7851     18181



In [18]:
# confusion matrix
cm = confusion_matrix(y_test, predict)
print("Akurasi Prediksi : ",np.round(accuracy_score(y_test, predict) * 100),"%")

print("Hasil dari Confusion Matrix")
print(cm)


Akurasi Prediksi :  81.0 %
Hasil dari Confusion Matrix
[[13616   626]
 [ 2767  1172]]


# Eksperimen Menggunakan KNN

Pada tahap eksperimen ini dilakukan dengan metode **KNN** untuk mengetahui apakah prediksi dari akurasi model lebih tinggi dari **Logistic Regression**

In [19]:
knn_x_test = df_salju_test.iloc[1:, :-1].values
knn_y_test = df_salju_test.iloc[1:, -1].values

knn_x_train = df_salju_train.iloc[1:, :-1].values
knn_y_train = df_salju_train.iloc[1:, -1].values

In [20]:
scaler = MinMaxScaler()
knn_x_train = scaler.fit_transform(x_train)
knn_x_test = scaler.transform(x_test)

Memanggil Library KNN untuk mengklasifikasi data

In [21]:
from sklearn.neighbors import KNeighborsClassifier
model_knn = KNeighborsClassifier(n_neighbors = 5)
model_knn.fit(knn_x_train, knn_y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [22]:
pred_knn = model_knn.predict(knn_x_test)
result_knn = pd.concat([pd.DataFrame(knn_y_test, columns=['Pasti']), pd.DataFrame(pred_knn, columns=['Prediksi'])], axis=1)
result_knn

Pasti  Prediksi
0          1         0
1          0         0
2          0         0
3          0         1
4          0         0
...      ...       ...
18176      0         0
18177      0         0
18178      0         0
18179      0         0
18180      1         0

[18181 rows x 2 columns]

In [23]:
print(classification_report(knn_y_test, pred_knn, digits=4))

              precision    recall  f1-score   support

           0     0.8430    0.9186    0.8791     14242
           1     0.5642    0.3813    0.4551      3939

    accuracy                         0.8022     18181
   macro avg     0.7036    0.6499    0.6671     18181
weighted avg     0.7826    0.8022    0.7873     18181



In [24]:
# confusion matrix
cm = confusion_matrix(knn_y_test, pred_knn)
print("Akurasi Prediksi : ",format (np.round(accuracy_score(knn_y_test, pred_knn) * 100)),"%")

print("Hasil dari Confusion Matrix")
print(cm)

Akurasi Prediksi :  80.0 %
Hasil dari Confusion Matrix
[[13082  1160]
 [ 2437  1502]]


# Kesimpulan

Klasifikasi utama menggunakan Logistic Regression diperoleh akurasi sebesar **0.81%** 


Pada eksperimen pertama dengan menggunakan metode KNN diperoleh hasil prediksi sebesar **0.80%**

Sehingga dapat simpulkan bahwa metode klasifikasi yang tertinggi untuk akurasi adalah metode Logistic Regression